File: src/hyper_markdown/parse.py
"""Source bytes -> `Document` (HMD-0001 §1, §2, §3)."""

from __future__ import annotations

import re
from pathlib import Path

from markdown.extensions.toc import slugify, unique

from . import scan
from .frontmatter import parse_card_config, parse_yaml, split_frontmatter
from .model import (
    ERROR,
    Anchor,
    CardConfig,
    Diagnostic,
    Document,
    Heading,
    Link,
    Span,
)

#: Characters that may not appear literally in a link target (HMD-0001 §2).
RESERVED_CHARS = frozenset("[]|#^\n")

_BLOCK_ID_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_-]{0,63}$")


def slug_for(text: str, used: set[str]) -> str:
    """Assign a heading slug the way Python-Markdown's `toc` extension does.

    Delegated rather than invented: a link and its rendered destination cannot
    disagree if both sides use the same rule (HMD-0001 §3).
    """
    return unique(slugify(text, "-"), used)


def parse(path: Path, text: str, rel: str) -> Document:
    """Parse one `.hmd` file. `rel` is the root-relative path for diagnostics."""
    diagnostics: list[Diagnostic] = []

    raw_yaml, _body, body_offset = split_frontmatter(text)
    frontmatter: dict = {}
    card = CardConfig()

    if raw_yaml is not None:
        try:
            frontmatter = parse_yaml(raw_yaml)
        except ValueError as exc:
            diagnostics.append(Diagnostic("HMD009", ERROR, rel, 1, 1, str(exc)))
        else:
            card, problems = parse_card_config(frontmatter)
            for rule, message in problems:
                diagnostics.append(Diagnostic(rule, ERROR, rel, 1, 1, message))

    # Mask the body only; frontmatter is YAML, not markdown.
    masked = " " * body_offset + scan.mask(text[body_offset:])

    headings = _headings(text, masked)
    anchors = _anchors(text, masked)
    links, link_diagnostics = _links(text, masked, rel)
    diagnostics.extend(link_diagnostics)

    return Document(
        path=path,
        text=text,
        frontmatter=frontmatter,
        card=card,
        headings=headings,
        anchors=anchors,
        links=links,
        diagnostics=diagnostics,
    )


def _span(text: str, start: int, end: int) -> Span:
    line, column = scan.line_col(text, start)
    return Span(start=start, end=end, line=line, column=column)


def _headings(text: str, masked: str) -> tuple[Heading, ...]:
    used: set[str] = set()
    out = []
    for level, heading_text, start, end in scan.find_headings(masked):
        # Slug from the ORIGINAL text: masking blanks inline code, and a heading
        # such as `## The \`hmd\` CLI` must slug from what the renderer sees.
        original = text[start:end]
        m = re.match(r"^[ \t]{0,3}#{1,6}[ \t]+(.*?)[ \t]*$", original)
        display = m.group(1) if m else heading_text
        display = re.sub(r"[ \t]+#+[ \t]*$", "", display)
        display = re.sub(r"[ \t]+\^[A-Za-z0-9][A-Za-z0-9_-]{0,63}[ \t]*$", "", display).strip()
        out.append(
            Heading(level=level, text=display, slug=slug_for(display, used), span=_span(text, start, end))
        )
    return tuple(out)


def _anchors(text: str, masked: str) -> tuple[Anchor, ...]:
    return tuple(
        Anchor(block_id=block_id, span=_span(text, start, end))
        for block_id, start, end in scan.find_anchors(masked)
    )


def _links(text: str, masked: str, rel: str) -> tuple[tuple[Link, ...], list[Diagnostic]]:
    links: list[Link] = []
    diagnostics: list[Diagnostic] = []

    for is_embed, inner, start, end in scan.find_links(masked):
        span = _span(text, start, end)
        raw = text[start:end]
        try:
            page_ref, fragment, fragment_kind, display = _parse_target(inner)
        except ValueError as exc:
            diagnostics.append(Diagnostic("HMD010", ERROR, rel, span.line, span.column, f"{exc} in {raw}"))
            continue
        links.append(
            Link(
                page_ref=page_ref,
                fragment=fragment,
                fragment_kind=fragment_kind,
                display=display,
                is_embed=is_embed,
                raw=raw,
                span=span,
            )
        )

    for start in scan.find_unterminated(masked):
        line, column = scan.line_col(text, start)
        diagnostics.append(Diagnostic("HMD010", ERROR, rel, line, column, "unterminated '[[' — expected ']]'"))

    return tuple(links), diagnostics


def _parse_target(inner: str) -> tuple[str, str | None, str | None, str | None]:
    """Split a link body into (page_ref, fragment, fragment_kind, display).

    Raises `ValueError` for anything malformed (HMD010).
    """
    if not inner.strip():
        raise ValueError("empty link target")

    # Display text comes last and may itself contain '|', so split once only.
    display: str | None = None
    if "|" in inner:
        inner, display = inner.split("|", 1)
        if not display.strip():
            raise ValueError("empty display text")

    fragment: str | None = None
    fragment_kind: str | None = None
    if "#" in inner:
        inner, fragment = inner.split("#", 1)
        if fragment.startswith("^"):
            fragment_kind, fragment = "block", fragment[1:]
            if not _BLOCK_ID_RE.match(fragment):
                raise ValueError(f"invalid block id {fragment!r}")
        else:
            fragment_kind = "heading"
            if not fragment.strip():
                raise ValueError("empty heading fragment")

    page_ref = inner.strip()
    if not page_ref:
        # Covers `[[#tag]]`: a tag is never a link target, because '#' is
        # reserved for fragments (HMD-0001 §5.3).
        raise ValueError("link target has no page reference")

    bad = RESERVED_CHARS.intersection(page_ref)
    if bad:
        raise ValueError(f"reserved character {''.join(sorted(bad))!r} in target")

    return page_ref, fragment, fragment_kind, display

File: tools/hmd-ts-core/src/parse.ts
/**
 * Source text -> `ParsedDocument` (HMD-0001 §1, §2, §3) — a port of
 * `src/hyper_markdown/parse.py`.
 *
 * Parsing never throws on partial input (HMD-0020 §7). A buffer caught
 * mid-keystroke is the ordinary input here, not an error case: an unterminated
 * construct becomes a diagnostic and the rest of the card parses normally.
 */

import { parseCardConfig, parseYaml, splitFrontmatter } from "./frontmatter.js";
import {
  ERROR,
  emptyCardConfig,
  type Anchor,
  type CardConfig,
  type Diagnostic,
  type FragmentKind,
  type Heading,
  type Link,
  type ParsedDocument,
  type Span,
} from "./model.js";
import * as scan from "./scan.js";
import { slugFor } from "./slug.js";

export const SUFFIX = ".hmd";

/** Characters that may not appear literally in a link target (HMD-0001 §2). */
export const RESERVED_CHARS: ReadonlySet<string> = new Set(["[", "]", "|", "#", "^", "\n"]);

const BLOCK_ID_RE = /^[A-Za-z0-9][A-Za-z0-9_-]{0,63}$/;
const HEADING_TEXT_RE = /^[ \t]{0,3}#{1,6}[ \t]+([\s\S]*?)[ \t]*$/;
const CLOSING_HASHES_RE = /[ \t]+#+[ \t]*$/;
const TRAILING_ANCHOR_RE = /[ \t]+\^[A-Za-z0-9][A-Za-z0-9_-]{0,63}[ \t]*$/;

/** Parse one `.hmd` file. `rel` is the root-relative path used in diagnostics. */
export function parse(rel: string, text: string): ParsedDocument {
  const diagnostics: Diagnostic[] = [];
  const { raw, bodyOffset } = splitFrontmatter(text);

  let frontmatter: Record<string, unknown> = {};
  let card: CardConfig = emptyCardConfig();

  if (raw !== null) {
    try {
      frontmatter = parseYaml(raw);
      const parsed = parseCardConfig(frontmatter);
      card = parsed.card;
      for (const [rule, message] of parsed.problems) {
        diagnostics.push({ rule, severity: ERROR, path: rel, line: 1, column: 1, message });
      }
    } catch (exc) {
      diagnostics.push({
        rule: "HMD009",
        severity: ERROR,
        path: rel,
        line: 1,
        column: 1,
        message: exc instanceof Error ? exc.message : String(exc),
      });
    }
  }

  // Mask the body only; frontmatter is YAML, not markdown. The prefix is
  // spaces rather than the original bytes, exactly as the canonical
  // implementation writes it — see the note on `maskedBody` below.
  const masked = " ".repeat(bodyOffset) + scan.mask(text.slice(bodyOffset));
  const index = new scan.PositionIndex(text);

  const headings = collectHeadings(text, masked, index);
  const anchors = collectAnchors(text, masked, index);
  const { links, diagnostics: linkDiagnostics } = collectLinks(text, masked, index, rel);
  diagnostics.push(...linkDiagnostics);

  return { path: rel, text, frontmatter, card, headings, anchors, links, diagnostics };
}

function spanAt(index: scan.PositionIndex, start: number, end: number): Span {
  const { line, column } = index.at(start);
  return { start, end, line, column };
}

function collectHeadings(
  text: string,
  masked: string,
  index: scan.PositionIndex,
): Heading[] {
  const used = new Set<string>();
  const out: Heading[] = [];
  for (const raw of scan.findHeadings(masked)) {
    // Slug from the ORIGINAL text: masking blanks inline code, and a heading
    // such as "## The `hmd` CLI" must slug from what the renderer sees.
    const original = text.slice(raw.start, raw.end);
    const m = HEADING_TEXT_RE.exec(original);
    let display = m ? m[1]! : raw.text;
    display = display.replace(CLOSING_HASHES_RE, "");
    display = display.replace(TRAILING_ANCHOR_RE, "").trim();
    out.push({
      level: raw.level,
      text: display,
      slug: slugFor(display, used),
      span: spanAt(index, raw.start, raw.end),
    });
  }
  return out;
}

function collectAnchors(text: string, masked: string, index: scan.PositionIndex): Anchor[] {
  return scan
    .findAnchors(masked)
    .map((raw) => ({ blockId: raw.blockId, span: spanAt(index, raw.start, raw.end) }));
}

function collectLinks(
  text: string,
  masked: string,
  index: scan.PositionIndex,
  rel: string,
): { links: Link[]; diagnostics: Diagnostic[] } {
  const links: Link[] = [];
  const diagnostics: Diagnostic[] = [];

  for (const found of scan.findLinks(masked)) {
    const span = spanAt(index, found.start, found.end);
    const raw = text.slice(found.start, found.end);
    let target: ParsedTarget;
    try {
      target = parseTarget(found.inner);
    } catch (exc) {
      diagnostics.push({
        rule: "HMD010",
        severity: ERROR,
        path: rel,
        line: span.line,
        column: span.column,
        message: `${exc instanceof Error ? exc.message : String(exc)} in ${raw}`,
      });
      continue;
    }
    links.push({ ...target, isEmbed: found.isEmbed, raw, span });
  }

  for (const start of scan.findUnterminated(masked)) {
    const { line, column } = index.at(start);
    diagnostics.push({
      rule: "HMD010",
      severity: ERROR,
      path: rel,
      line,
      column,
      message: "unterminated '[[' — expected ']]'",
    });
  }

  return { links, diagnostics };
}

interface ParsedTarget {
  pageRef: string;
  fragment: string | null;
  fragmentKind: FragmentKind | null;
  display: string | null;
}

/**
 * Split a link body into its parts, throwing for anything malformed (HMD010).
 */
export function parseTarget(inner: string): ParsedTarget {
  if (inner.trim() === "") throw new Error("empty link target");

  // Display text comes last and may itself contain '|', so split once only.
  let rest = inner;
  let display: string | null = null;
  const bar = rest.indexOf("|");
  if (bar !== -1) {
    display = rest.slice(bar + 1);
    rest = rest.slice(0, bar);
    if (display.trim() === "") throw new Error("empty display text");
  }

  let fragment: string | null = null;
  let fragmentKind: FragmentKind | null = null;
  const hash = rest.indexOf("#");
  if (hash !== -1) {
    fragment = rest.slice(hash + 1);
    rest = rest.slice(0, hash);
    if (fragment.startsWith("^")) {
      fragmentKind = "block";
      fragment = fragment.slice(1);
      if (!BLOCK_ID_RE.test(fragment)) {
        throw new Error(`invalid block id '${fragment}'`);
      }
    } else {
      fragmentKind = "heading";
      if (fragment.trim() === "") throw new Error("empty heading fragment");
    }
  }

  const pageRef = rest.trim();
  if (pageRef === "") {
    // Covers `[[#tag]]`: a tag is never a link target, because '#' is
    // reserved for fragments (HMD-0001 §5.3).
    throw new Error("link target has no page reference");
  }

  const bad = [...new Set([...pageRef].filter((c) => RESERVED_CHARS.has(c)))].sort();
  if (bad.length > 0) {
    throw new Error(`reserved character '${bad.join("")}' in target`);
  }

  return { pageRef, fragment, fragmentKind, display };
}
